### PR TITLE
Add support of libFuzzer

### PR DESCRIPTION
**Pull request by mmisono**
[LibFuzzer](https://llvm.org/docs/LibFuzzer.html) is a coverage-guided fuzzer developed with llvm/clang.
LibFuzzer's main target is a function, and the biggest difference
between AFL and the libfuzzer is that the former re-executes the program
after each execution (actually it does fork), while the libFuzzer keeps
calling the target function in the main loop.

The main target of the libfuzzer is a function, but with a few
changes, it can be fuzzed from the beginning of the program, like the
AFL.

Because the current bpftrace is known to have memory leaks and there
might be probably some global states, it may not be suitable for
libfuzzer. But libFuzzer already found several bugs (#1618, #1621). Also,
it seems oss-fuzz expects the libfuzzer interface, so support it is a good thing.

##### Checklist

- [ ] Language changes are updated in `docs/reference_guide.md`
- [ ] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [ ] The new behaviour is covered by tests

### CODE REVIEW

File: src/ast/field_analyser.cpp [hunk 1]
```diff
@@ -294,7 +294,11 @@ int FieldAnalyser::analyse()
   std::string errors = err_.str();
   if (!errors.empty())
   {
+#ifdef FUZZ
+    (void)out_;
```
Comment (fbs) [inline src/ast/field_analyser.cpp:298]: Both the field analyser and semantic allow setting `out` in the constructor. Setting it to `/dev/null` from fuzz_main.cpp seems cleaner than doing ifdefs everywhere

Comment (mmisono) [reply src/ast/field_analyser.cpp]: I didn't notice it. Thanks, fixed it.

File: src/log.cpp [hunk 2]
```diff
@@ -196,7 +196,16 @@ LogStream::LogStream(const std::string& file,
 
 LogStream::~LogStream()
 {
+#ifdef FUZZ
+  // When fuzzing, we don't want to output error messages. However, some
```
Comment (danobi) [inline src/log.cpp:200]: Why do we need to suppress error messages? Is it a speed thing? I thought fuzzing engines look for program crashes -- the output shouldn't matter to them.

Comment (mmisono) [reply src/log.cpp:200]: libfuzzer prints the program's stdout/stderr to a terminal by default, and it makes reading libfuzzer's output almost impossible. We can change the behavior `-close_fd_mask` flags of libfuzzer( https://llvm.org/docs/LibFuzzer.html). But I think suppressing in bpftrace is better. And maybe another fuzzer behaves like libfuzzer.

File: docs/fuzzing.md [hunk 3]
```diff
@@ -104,6 +104,29 @@ I describe several important things:
...
+and bpftrace can be fuzzed with it.
+
+### Compile with libFuzzer
+To use the libFuzzer, use clang and compile it as follows.
+
+```
+CC=clang-10 CXX=clang++-10 cmake .. -DBUILD_ASAN=1 -DBUILD_FUZZ=1 -DUSE_LIBFUZZER=1 -DBUILD_TESTING=0
+```
+
+`-DBUILD_FUZZ=1` and `-DUSE_LIBFUZZE=1` is necessary.
```
Comment (mmisono) [inline docs/fuzzing.md:118]: ah, I'll fix typo

File: src/log.cpp (hunk 2 above)
Comment (danobi) [reply src/log.cpp:200]: Ah, got it.

File: CMakeLists.txt [hunk 4]
```diff
@@ -42,14 +43,23 @@ endif()
...
-    message(STATUS "Fuzzing seantic analyzer")
-    add_compile_options("-DTEST_SEMANTIC")
-elseif(FUZZ_TARGET STREQUAL "codegen")
-    message(STATUS "Fuzzing codegen")
-else()
-    message(FATAL_ERROR "Unsupported FUZZ_TARGET")
+
+    if(FUZZ_TARGET STREQUAL "semantic")
+        message(STATUS "Fuzzing seantic analyzer")
+        add_compile_options("-DTEST_SEMANTIC")
```
Comment (danobi) [inline CMakeLists.txt:49]: I think `target_compile_options()` is better b/c it doesn't proliferate the macro everywhere. Also probably more portable.

**Review by fbs — APPROVED** [review]

File: CMakeLists.txt (hunk 4 above)
Comment (mmisono) [reply CMakeLists.txt]: I didn't know that option. Now I move to fuzz related settings to src/CMakeList.txt and use it.

Comment (mmisono) [reply CMakeLists.txt]: ah, target_compile_definitions looks better.

Comment (mmisono) [reply CMakeLists.txt]: hmm.. I tried to use target_compile_definitions but it does not work the same as the target_compile_options (encountered linking error with libfuzzer.a). CMake is difficult... I'll keep the current code.

**Pull request merged**